service/matching/task.go

```diff
@@ -160,6 +160,7 @@ func (task *internalTask) finish(err error) {
 	case task.responseC != nil:
 		task.responseC <- err
 	case task.event.completionFunc != nil:
+		// TODO: this probably should not be done synchronosly in PollWorkflow/ActivityTaskQueue
```

Comment:
If not done synchronously, the task might be dispatched again

Reply:
I am not following. The task has already been removed from the task buffer and later poller won't be able to see it before it's persisted/loaded again. 
And the persistence part can be handled by a background component?

Reply:
Hmmm, I'm not entirely sure. This call is to ack the task. If we don't do it before return, the ack could fail and task could be dispatched again. However, as you said that only happen after persisted/reloaded again from DB. And since the history already recored the task, so it will be deduped anyway. So I agree that we can to this async.